pkg/archive/archive.go

```diff
@@ -530,7 +550,8 @@ func createTarFile(path, extractDir string, hdr *tar.Header, reader io.Reader, L
 		}
 
 	case tar.TypeLink:
-		targetPath := filepath.Join(extractDir, hdr.Linkname)
+		lpath := adjustHardLink(hdr.Linkname, hdr.Name)
+		targetPath := filepath.Join(extractDir, lpath)
```

Comment:
@jlhawn 

> One alternative that I have warmed up to is that if the source of the archive is a directory, then create the archive by going into that source directory and archiving . such that all archive entries begin with . and then the extract logic would be:

Yup, that's exactly how I would want to do it. Effectively, make the target `filepath.Dir(target)` if it's a file and just `target` if it's a directory.

> I may be overlooking something, but it should be fine as long as all of the existing CLI tests around docker cp still pass.

I'm personally a little more cautious (given how many different breakouts I've found with `docker cp` in the past). If we can do it with `filepath` we should.